### DISCUSSION TITLE
Accelerate time in grafana

### DISCUSSION THREAD

**Discussion by dincerunal** [grafana/grafana#38226] [Dashboards, Panels & Visualizations]
Hello,

**What would you like to be added**:
I want to speed up the processing time of Grafana at x5, x10, x20, x50, x100 ratios for my test processes.

**Why is this needed**:

Actually what I want to do,

I created dashboards to monitor the alert status in grafana. I created fake data in my system to simulate my alert situations on these boards. The time of this data covers the range now, now + 12h. In fact, it takes a long time to analyze the alert status in real data. For this reason, I cannot be very flexible on my alert rules. I have to wait until the end of this period to see alert status in the system. (I have many states like this actually.) Grafana creates pending, alerting, and ok states according to the records in my database.  I want to accelerate Grafana's processing time at x5, x10, x20, x50, x100 ratios, and to perform this process in a much shorter time in real time.

**Comment by torkelo** [comment]
@kylebrandt does this make sense? Is any way doable in practice?

**Comment by kylebrandt** [comment]
This sounds like the "historical testing" feature that is in Bosun, that would output something like:

The main problem is that it is fairly expensive to do in a data source agnostic way. The way worked in Bosun is you would select a time range, and then an interval or a number of queries to run.

> Setting both From and To enables testing multiple iterations of the selected alert over time. The number of iterations depends on the setting to the two linked fields Intervals and Step Duration at ③. Changing one changes the other. Intervals will be the number of runs to do even spaced out over the duration of From to To and Step Duration is how much time in minutes should be between intervals. Doing a test over time will populate the Timeline tab ⑤ which draws a clickable graphic of severity states for each item in the set:

![image](https://user-images.githubusercontent.com/1692624/109059349-290e6e80-76b2-11eb-98ac-3b265d25fef4.png)

> Setting both From and To enables testing multiple iterations of the selected alert over time. The number of iterations depends on the setting to the two linked fields Intervals and Step Duration at ③. Changing one changes the other. Intervals will be the number of runs to do even spaced out over the duration of From to To and Step Duration is how much time in minutes should be between intervals. Doing a test over time will populate the Timeline tab ⑤ which draws a clickable graphic of severity states for each item in the set:

It would then run all those queries with a pool limiting simultaneous queries. For an interval of say 5 minutes, it would run adjacent 5 minute queries.

So this would speed up the alert authoring and testing workflow significantly. But it would best be implemented as a job system. This is because with more expensive queries, or range/interval combination that is a fair amount of runs, it may take a minute or so - so having to wait on an open network connection is less ideal.

So I found I generally used in two modes:
 - To tweak a specific alert that had fired at some time
 - To get a general overview of how much the alert rule would trigger for the historical data

For the general over, a larger time range is generally wanted, which means more queries if the interval is kept the same. And with a feature like FOR (Pending), you would have to use the same interval it would actually run at.

So possible, has some limitations, and some care needs to be taken to do it right. But extremely useful in my experience.

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Comment by dincerunal** [comment]
Hi everyone, 

First of all, let me talk about our architecture of using grafana in the system. In my system, there are more than 20 nodes whose metrics I monitor, and about 120 applications are running on each node. I monitor system metrics such as cpu and memory of these nodes and applications. In addition, there is a mechanism that I developed for these applications to communicate. I am also examining and collecting the statistics of this mechanism. (There are 17 different metrics that we are currently monitoring in total.) There is an agent that I have developed and working on each node to collect these metrics. This agent collects these 17 metrics and writes them to the mysql database we have determined, and we view this data with grafana. So far everything is going as we want. To make these metrics meaningful, I set some threshold values ​​using grafana's alert system. To give an example of my Alert rules, if an application's memory usage increases for 4 hours, it indicates an alert for my case. This and more are available on my alert thresholds with larger timeframes. I don't have a problem with this part. The system generally works in this structure.

Then, if we talk about what I want to do, to test this system, I am writing 4 hours of data to the mysql database, the first record of which is now and prospectively. (now to now+4h). I have to wait 4 hours to confirm this alert in Grafana. Instead of waiting this long to verify my tests, I can test grafana's running speed at x2, x4, x8, x16, x32 etc. If I can configure it as, this waiting time will be much less. If I can sort of set the time of the grafana world to flow faster than real time, my problem will be solved.

Thank you for your interest in advance as I met a successful and enthusiastic team. I'm waiting for your help. :)

**Comment by yuri-tceretian** [comment]
I agree with Kyle that in general case testing rule on an arbitrary interval of time would require some job-like approach. Unfortunately, there is no such feature in the Grafana code. So we will need to create it from the scratch. Also, Grafana is meant to be stateless (in the future) and therefore bringing support for jobs would violate its statelessness. 

As a first approach, we decided to implement a quick POC. See PR PoC https://github.com/grafana/grafana/pull/57318 that will support testing alert on an interval of time.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
